### PR TITLE
ADD: Image upload functionality

### PR DESCRIPTION
**Pull request by Uzay-G**
random

Stores image in a `images` subdirectory inside the user directory.

### CODE REVIEW

**Comment by Uzay-G** [comment]
@srevinsaju could you test this out? Thanks :smile:

File: archivy/data.py [hunk 1]
```diff
@@ -314,3 +314,35 @@ def open_file(path):
...
+    dest_path = base_path / f"{sanitized_filename}.{fileparts[1]}"
+    i = 1
+    while dest_path.exists():
+        dest_path = base_path / f"{sanitized_filename}-{i}.{fileparts[1]}"
+        i += 1
+    image.save(str(dest_path))
+    return dest_path.parts[-1]
+
+
+def image_exists(filename):
```
Comment (srevinsaju) [inline archivy/data.py:342]: Can we add a type hint over here, it looks like it can be both an `int` or a `str`.

File: archivy/api.py [hunk 2]
```diff
@@ -192,3 +192,19 @@ def search_endpoint():
...
+        return jsonify({"error": "400"}), 400
+    image = request.files["image"]
+    if (
+        data.valid_image_filename(image.filename)
+        and image.headers["Content-Type"].strip() in CONTENT_TYPES
+    ):
+        saved_to = data.save_image(image)
+        return jsonify({"data": {"filePath": f"/images/{saved_to}"}}), 200
+    else:
+        return jsonify({"error": "415"}), 415
```
Comment (srevinsaju) [inline archivy/api.py:210]: These two can be compressed to just
```python
    return jsonify({"error": "415"}), 415
```

File: archivy/data.py [hunk 3]
```diff
@@ -314,3 +314,35 @@ def open_file(path):
...
+    sanitized = secure_filename(filename)
+    image_path = Path(current_app.config["USER_DIR"]) / "images" / sanitized
+    if image_path.exists():
+        return str(image_path)
+    else:
+        return 0
```
Comment (srevinsaju) [inline archivy/data.py:348]: This can be compressed to 
```python
    return 0
```

**Review by srevinsaju — CHANGES_REQUESTED** [review]
Haven't completely reviewed this yet, but I will look into this PR in a while.

**Comment by Uzay-G** [comment]
Ok, I implemented all your suggestions :)

**Comment by srevinsaju** [comment]
I tried using the Bookmark feature, is it possible to archive / save all of those images to an image file instead of using a link. It might be harder, but it would be a great step to good archive. What do you think?

**Comment by Uzay-G** [comment]
Oh so you mean bookmarking a link to an image would save it to the images file?

That could be interesting (after this is merged though). Has the image upload function been working though?

**Comment by srevinsaju** [comment]
> Oh so you mean bookmarking a link to an image would save it to the images directory?


Well, I thought this included this feature too, but what I meant was iterating through all images in a html webpage while scraping and downloading all these images, and saving them in `images/` folder so that these images are still accessible

> 
> That could be interesting (after this is merged though). Has the image upload function been working though?

I just checked out this PR once again. The image upload feature isn't perfect yet. Few things to consider:
* In case the file is huge, the image upload takes a lot of time. This makes the user think that it is a bug. Instead, a progress bar or busy indicator is highly required.
* When trying to upload the image, the `` markup does not appear immediately. When trying to upload the same image twice, `` doesn't appear immediately, unless this is attempted many times. On the `nth` time, n \* `` appears on archivy

image

The last line looks like a bug. The same number (i.e. 2) is repeated twice in the file name.

**Comment by Uzay-G** [comment]
It's not very obvious, but there is actually a progress bar at the bottom (when you upload): 
image

I haven't had any problems with the speed of uploads :thinking:, especially since the limit (that the editor lib sets by default) is at 2 MB. By the way, do you think I should make that higher / remove it?

For the issue with the  markup being added, I think that just happens if you really mass upload images and the bug you mentioned must just be a timing issue with when images are being saved, in that case.

If everything else works with you @srevinsaju, I'll merge this in and I think I'll add a config option in another PR to save all images found in webpages, like you said.

**Comment by srevinsaju** [comment]
> It's not very obvious, but there is actually a progress bar at the bottom (when you upload):

Oh, didn't notice that one.

> 
> 
> I haven't had any problems with the speed of uploads , especially since the limit (that the editor lib sets by default) is at 2 MB. By the way, do you think I should make that higher / remove it?

I guess you should remove it, or add a configuration entry to archivy's configuration and let the user keep the cap on the file size.  Preferably setting the default to `-1`, indicating no limit.

> 
> For the issue with the  markup being added, I think that just happens if you really mass upload images and the bug you mentioned must just be a timing issue with when images are being saved, in that case.

This only happens when I try to add the same image multiple times into the editor. Using different filenames does not create this bug.

> 
> If everything else works with you @srevinsaju, I'll merge this in and I think I'll add a config option in another PR to save all images found in webpages, like you said.

I guess its fine.

**Comment by Uzay-G** [comment]
> I guess you should remove it, or add a configuration entry to archivy's configuration and let the user keep the cap on the file size. Preferably setting the default to -1, indicating no limit.

Good idea.

> This only happens when I try to add the same image multiple times into the editor. Using different filenames does not create this bug.

Yeah, I think there's a timing issue when files with identical filenames are saved repeatedly :thinking:

**Pull request merged**